orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/DetermineHealthProvidersTask.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.clouddriver.tasks;
+
+import com.google.common.collect.ImmutableMap;
+import com.netflix.frigga.Names;
+import com.netflix.spinnaker.orca.DefaultTaskResult;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.RetryableTask;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.clouddriver.utils.CloudProviderAware;
+import com.netflix.spinnaker.orca.front50.Front50Service;
+import com.netflix.spinnaker.orca.front50.model.Application;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+@Component
+public class DetermineHealthProvidersTask implements RetryableTask, CloudProviderAware {
+  private static final Logger log = LoggerFactory.getLogger(DetermineHealthProvidersTask.class);
+  private static final Map<String, String> HEALTH_PROVIDER_NAMES_BY_PLATFORM = ImmutableMap
+    .<String, String>builder()
+    .put("aws", "Amazon")
+    .put("gce", "Google")
+    .put("openstack", "Openstack")
+    .put("titus", "Titus")
+    .build();
+
+  @Autowired
+  Front50Service front50Service;
+
+  @Override
+  public TaskResult execute(Stage stage) {
+    if (stage.getContext().containsKey("interestingHealthProviderNames")) {
+      // should not override any stage-specified health providers
+      return new DefaultTaskResult(ExecutionStatus.SUCCEEDED);
+    }
+
+    String platformSpecificHealthProviderName = HEALTH_PROVIDER_NAMES_BY_PLATFORM.get(getCloudProvider(stage));
+    if (platformSpecificHealthProviderName == null) {
+      log.warn("Unable to determine platform health provider for unknown cloud provider '{}'", getCloudProvider(stage));
+      return new DefaultTaskResult(ExecutionStatus.SUCCEEDED);
+    }
+
+    try {
+      String applicationName = (String) stage.getContext().get("application");
+      if (applicationName == null && stage.getContext().containsKey("serverGroupName")) {
```

Comment:
This was sufficient to cover the various scenarios that I tested ... did I miss anything obvious?